libc/src/__support/threads/linux/rwlock.h

```diff
@@ -0,0 +1,412 @@
+    return (state & MASK) < MAX_READERS && !has_readers_waiting(state) &&
+           !has_writers_waiting(state);
+  }
+  LIBC_INLINE static constexpr bool
+  has_reached_max_readers(FutexWordType state) {
+    return (state & MASK) == MAX_READERS;
+  }
+
+  // Convert a relative timeout to an absolute timespec.
+  LIBC_INLINE static void abs_timeout(cpp::optional<timespec> &timeout) {
+    if (!timeout)
+      return;
+
+    int errno_backup = libc_errno;
+    timespec now;
+    // if we failed to get time, we move timeout to infinity by setting it to
+    // nullopt.
+    if (LIBC_NAMESPACE::clock_gettime(CLOCK_MONOTONIC, &now) != 0) {
+      timeout = cpp::nullopt;
+      libc_errno = errno_backup;
+      return;
+    }
+    if (__builtin_add_overflow(now.tv_sec, timeout->tv_sec, &timeout->tv_sec)) {
+      timeout = cpp::nullopt;
+      return;
+    }
+    if (__builtin_add_overflow(now.tv_nsec, timeout->tv_nsec,
+                               &timeout->tv_nsec)) {
+      timeout = cpp::nullopt;
+      return;
+    }
+    if (now.tv_nsec >= 1'000'000'000) {
+      if (__builtin_add_overflow(now.tv_sec, 1, &timeout->tv_sec)) {
+        timeout = cpp::nullopt;
+        return;
+      }
+      timeout->tv_nsec -= 1'000'000'000;
+    }
+  }
+
+  template <typename F>
+  LIBC_INLINE bool fetch_update(cpp::Atomic<FutexWordType> &__restrict word,
+                                FutexWordType &__restrict prev,
+                                cpp::MemoryOrder set_order,
+                                cpp::MemoryOrder fetch_order, F &&f) {
+    prev = word.load(fetch_order);
+    // It is okay to have spurious failures here as we are in a loop.
+    while (cpp::optional<FutexWordType> new_val = f(prev))
+      if (word.compare_exchange_weak(prev, *new_val, set_order, fetch_order))
+        return true;
+
+    return false;
+  }
+
+  template <typename F>
+  LIBC_INLINE FutexWordType spin_until(int_fast32_t spin, F &&f) {
+    for (;;) {
+      FutexWordType state = this->state.load(cpp::MemoryOrder::RELAXED);
+      if (f(state) || spin == 0)
+        return state;
+      // pause the pipeline to avoid extra memory loads due to speculation
+      sleep_briefly();
+      --spin;
+    }
+  }
+
+  LIBC_INLINE static bool
+  is_timeout(const cpp::optional<timespec> &abs_timeout) {
+    if (!abs_timeout)
+      return false;
+    timespec now;
+    LIBC_NAMESPACE::clock_gettime(CLOCK_MONOTONIC, &now);
+    if (now.tv_sec > abs_timeout->tv_sec ||
+        (now.tv_sec == abs_timeout->tv_sec &&
+         now.tv_nsec >= abs_timeout->tv_nsec))
+      return true;
+    return false;
+  }
+
+  LIBC_INLINE FutexWordType spin_read(int_fast32_t spin) {
+    // Stop spinning when it's unlocked or read locked, or when there's waiting
+    // threads.
+    return spin_until(spin, [&](FutexWordType state) -> bool {
+      return !is_write_locked(state) || has_readers_waiting(state) ||
+             has_writers_waiting(state) || is_unlocked(state);
```

Comment:
The `is_unlocked(state)` here seems redundant as `!is_write_locked(state)` implies either unlocked or read locked.